### DISCUSSION TITLE
Nodal Materials

### DISCUSSION THREAD

**Discussion by stephen-dixon** [idaholab/moose#24560] [Q&A General] [answered]
Can Functor materials be used as nodal materials? Otherwise are there any plans to implement nodal materials for use with the nodal kernel system? 

For context I'm currently trying to develop a reaction-diffusion type app with a large number of species, where the reaction rates at each point are calculated in an external library I'd like to treat as a black-box. I think I'd like to use nodal kernels and the AD system for this, but an initial implementation using a user object and aux-variables (similar to Thermochimica in the chemical reaction module) hasn't worked for me. I did manage to get something running with normal kernels and materials, but I'm concerned numerical stability could be an issue. 

Thanks for your time.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Not currently.
However, this is something we are considering because it would be very natural.
A new NodalArg for the functor spatial argument could be {node, element}  or {node, subdomain_id} and would naturally know which subdomain the material has been defined on.

Part of the problem with nodal materials has always been how to deal with different material definitions on different subdomains, as nodes can be at the interface between two subdomains. I think this is solved if we let the caller specify which subdomain the node is actually part of.

@lindsayad  will know more about issues with nodal materials

Guillaume

**Reply by lindsayad** [reply]
Hi @stephen-dixon thank you for your interest in functors 😄 Yes I think nodal functor evaluation could be on the way soon. See https://github.com/idaholab/moose/issues/19420 where adding nodal evaluations is a top priority. I agree with @GiudGiud that supplying a subdomain would alleviate the principal concern we've always had with "nodal materials", e.g. what happens at the intersection between multiple different materials. But by allowing the user to explicitly say what subdomain they want to associate with, that concern goes away

**Reply by stephen-dixon** [reply]
"Functors are great and we want more of them" love it. Thanks for the quick responses.
